src/matrix.jl

```diff
@@ -154,80 +154,81 @@ makemasks(matrix::AbstractArray) = makemask.(Ref(matrix), 1:8)
 Calculate a penalty score in order to pick the best mask.
 """
 function penalty(matrix::BitArray{2})
-    rows, cols = size(matrix)
-
+    n = size(matrix, 1)
     # Condition 1: 5+ in a row of the same color
-    score(c) = foldl(check, c, init = (0, !c[1], 0))[1]
-    function check((tot, pb, cnt), b)
-        pb != b && return (tot, b, 1)
-        cnt < 4 && return (tot, b, cnt + 1)
-        cnt == 4 && return (tot + 3, b, cnt + 1)
-        return (tot + 1, b, cnt)
+    function penalty1(line)
+        consecutive, score, cur = 0, 0, -1
+        for i in line
+            if i == cur
+                consecutive += 1
+            else
+                if consecutive ≥ 5
+                    score += consecutive - 2
+                end
+                consecutive, cur = 1, i
+            end
+        end
+        return score
     end
-    p1 = sum(map(score, eachrow(matrix))) + sum(map(score, eachcol(matrix)))
+    p1 = sum(penalty1, eachrow(matrix)) + sum(penalty1, eachcol(matrix))
 
     # Condition 2: number of 2x2 blocks of the same color
-    p2 = 0
-    for row in 1:rows - 1, col in 1:cols - 1
-        block = matrix[row:row + 1, col:col + 1]
-        if all(block) || all(.! block)
-            p2 += 2
-        end
-    end
+    p2 = 3 * count(matrix[i, j] == matrix[i+1, j] == matrix[i, j+1] ==
+                   matrix[i+1, j+1] for i in 1:n-1 for j in 1:n-1)
 
     # Condition 3: specific patterns in rows or columns
-    p3 = 0
     patt1 = BitArray([1, 0, 1, 1, 1, 0, 1, 0, 0, 0 ,0])
     patt2 = BitArray([0, 0, 0 ,0, 1, 0, 1, 1, 1, 0, 1])
-    for row in 1:rows, col in 1:cols - 10
-        hline = matrix[row, col:col + 10]
-        if all(.! xor.(hline, patt1)) || all(.! xor.(hline, patt2))
-            p3 += 40
-        end
-
-        vline = matrix[col:col + 10, row]
-        if all(.! xor.(vline, patt1)) || all(.! xor.(vline, patt2))
-            p3 += 40
-        end
+    function check(i, j)
+        hline = @view(matrix[i, j:j + 10])
+        vline = @view(matrix[j:j + 10, i])
+        return (hline == patt1 || hline == patt2) + (vline == patt1 || vline == patt2)
     end
-
+    p3 = 40 * sum(check(i, j) for i in 1:n for j in 1:(n-10))
```

Comment:
BTW, https://biojulia.net/post/hardware/ is a nice reading resource about hardware stuff on performance.

If we'd want to improve the performance here, we'd need to break the loop into smaller pieces and do finer-grained control on how to iterate through the memory efficiently. But it's better to do this kind of "dirty" optimization in future PRs.